NIFI-10975 Add Kubernetes Leader Election and State Provider

exceptionfactory opened the pull request:
# Summary

[NIFI-10975](https://issues.apache.org/jira/browse/NIFI-10975) Adds an initial implementation of Kubernetes cluster leader election and state management, supporting deployments without the need for ZooKeeper.

## State Management Implementation

The state management implementation uses Kubernetes [ConfigMaps](https://kubernetes.io/docs/concepts/configuration/configmap/) to persist cluster information for stateful components. The `KubernetesConfigMapStateProvider` uses the standard `data` property, but encodes property names using Base64 URL encoding without padding to meet [ConfigMap object](https://kubernetes.io/docs/concepts/configuration/configmap/) property naming requirements.

The new State Provider can be configured in `state-management.xml` using the following element definition, which is included in a commented section of the configuration:

```
<cluster-provider>
    <id>kubernetes-provider</id>
    <class>org.apache.nifi.kubernetes.state.provider.KubernetesConfigMapStateProvider</class>
</cluster-provider>
```

## Leader Election Implementation

The leader election implementation uses Kubernetes [Leases](https://kubernetes.io/docs/concepts/architecture/leases/) for distributed tracking of the current cluster leader. Kubernetes Lease names must adhere to [RFC 1123 subdomain naming requirements](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names), requiring a mapping from NiFi application names to Lease names.

The new Leader Election implementation can be configured using a new property in `nifi.properties` as follows:

```
nifi.cluster.leader.election.implementation=KubernetesLeaderElectionManager
```

## Framework Changes

The leader election implementation required promoting the `LeaderElectionManager` interface to `nifi-framework-api` to support NAR bundle extension loading. The `LeaderElectionManager` had two methods without runtime references, which were removed. Refactoring also involved creating a new `nifi-framework-leader-election-shared` module for abstracting tracking operations.

The `nifi.properties` configuration includes a new property with a default value of `CuratorLeaderElectionManager`, which provides current cluster coordination using ZooKeeper.

## Kubernetes Client

The implementation includes a new `nifi-kubernetes-client` library which provides utility components for Kubernetes Client access and service namespace determination.

The `nifi-kubernetes-client` library depends on the [Fabric8 Kubernetes Client](https://github.com/fabric8io/kubernetes-client) which supports current versions of Kubernetes and provides separation of API and implementation classes.

Both the State Provider and Leader Election Manager implementations attempt to resolve the Kubernetes namespace based on the standard [Service Account](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/) namespace secret. In absence of a readable namespace secret, the provider returns `default` as the namespace for storing Leases and ConfigMaps.

## Additional Changes

Additional changes include removing several integration test classes specific to ZooKeeper. These integration tests are less useful with current system integration tests run on a scheduled basis.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 8
  - [X] JDK 11
  - [X] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-kubernetes-bundle/nifi-framework-kubernetes-leader-election/src/main/java/org/apache/nifi/kubernetes/leader/election/KubernetesLeaderElectionManager.java:
```diff
@@ -0,0 +1,316 @@
+        return new StandardLeaderElectionCommandProvider(kubernetesClientProvider, namespace);
+    }
+
+    private void registerLeaderElectionCommand(final String roleName, final LeaderElectionStateChangeListener listener, final String participantId) {
```
There's a lot of check-then-modify here. I think we need to ensure that this is thread safe and synchronize the method or add locking

exceptionfactory replied:
This method is only called from `register`, which is marked as `synchronized`, do you think any additional synchronization is necessary?

markap14 replied:
Ah ok, good call. So technically it's correct/safe. But I would generally mark this as `synchronized` as well because it has virtually zero cost, since the synchronization lock has already been obtained, but it makes it more clear that the actions can only be taken while synchronized. Or alternatively documenting via JavaDoc that it should only be called while synchronized. But I'd prefer just synchronizing the method itself.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-kubernetes-bundle/nifi-framework-kubernetes-state-provider/src/main/java/org/apache/nifi/kubernetes/state/provider/KubernetesConfigMapStateProvider.java:
```diff
@@ -0,0 +1,317 @@
+        final ObjectMeta metadata = configMap.getMetadata();
+        final String resourceVersion = metadata.getResourceVersion();
+        try {
+            return resourceVersion == null ? UNKNOWN_VERSION : Long.parseLong(resourceVersion);
```
I don't believe this is valid. According to https://kubernetes.io/docs/reference/using-api/api-concepts/
```
You must not assume resource versions are numeric or collatable. API clients may only compare two resource versions for equality (this means that you must not compare resource versions for greater-than or less-than relationships).
```

markap14 replied:
I think I would propose that we update our StateMap as follows:
Make `long getVersion()` deprecated in favor of a new `String getStateVersion()`
Remove `long getVersion()` in version 2.0

exceptionfactory replied:
I was wondering about this approach, it probably depends on the Kubernetes implementation. One option that comes to mind is creating using a hash of the Resource Version. Another option could be using a custom metadata field.

markap14 replied:
Everywhere in the code base that I see `long getVersion()` being used, actually, is to check if the value is `-1` (indicating that the state hasn't actually been stored anywhere). So it probably actually makes sense to use an `Optional<String>` as the return type.

exceptionfactory replied:
Thanks for the suggestion, adding a new `Optional<String> getStateVersion()` method to `StateMap` sounds like a good approach.

exceptionfactory commented:
Thanks for the testing and feedback @markap14!

As it turns out, the Fabric8 Client version 6.3.0 includes a fix for relinquishing leadership described in [Issue 4547](https://github.com/fabric8io/kubernetes-client/issues/4547). I will include an update to client version 6.3.0 as part of additional updates for State Map versioning.

markap14 commented:
Great, thanks @exceptionfactory!
I also would recommend a couple minor changes to make this work more easily with Docker, as I had to make these changes to test:
In the `nifi-docker/dockerhub/sh/start.sh` file I added two lines:
```
# Set leader election and state management properties
prop_replace 'nifi.cluster.leader.election.implementation'      "${NIFI_LEADER_ELECTION_IMPLEMENTATION:-CuratorLeaderElectionManager}"
prop_replace 'nifi.state.management.provider.cluster'           "${NIFI_STATE_MANAGEMENT_CLUSTER_PROVIDER:-zk-provider}"
```

And also in the `nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/state-management.xml` file there's no need to add the `<cluster-provider>` element commented out. It's just configuration, it doesn't get used for anything unless it's selected in `nifi.properties` so we should just include the config uncommented in `state-management.xml` so that it can be easily referenced from nifi.properties

exceptionfactory commented:
Thanks for the recommendation on the Docker script @markap14. I pushed updates with the following changes:

1. Added `StateMap.getStateVersion()` returning `Optional<String>` and refactored all references
2. Upgraded Fabric8 Kubernetes Client from 6.2.0 to 6.3.0 to correct stop leading issue found during testing
3. Added `prop_replace` lines to Docker `start.sh` for setting the Leader Election and Cluster State Provider from environment variables

markap14 commented:
Thanks @exceptionfactory . Pulled the latest, rebuilt everything, rebuilt the docker image. Interestingly, it no longer worked. I got errors on startup telling me that the URI was invalid when calling `leaderElectionCommandProvider.findLeader(roleName);` (line 197 of KubernetesLeaderElectionManager) because the URI ended in "/Cluster Coordinator".
So I updated that code to get the `roleId` for the `roleName` and provide that.
That addressed the issue. It was now registering for the role properly.
Unfortunately, though, I still ran into the exact same issue. When I disconnected and reconnected a node, it gave the same stack trace, failing to release the lease because it was modified. This then caused the same issue, with both nodes thinking they are the leader.

markap14 commented:
Also, just to ensure that I was using the latest, I did a `kubectl exec -it nifi-0 bash` and went into the `work` directory to check the lib. It's using `kubernetes-client-6.3.1.jar` as expected.

exceptionfactory commented:
Thanks for the feedback @markap14.

On further evaluation of the disconnect and reconnect behavior, I realized the `unregister` method was not removing the local leader identifier from the `roleLeaders` Map within `KubernetesLeaderElectionManager`. The corresponding command was not being removed from the `roleCommands` Map, which was preventing proper registration on cluster reconnection. I corrected this behavior and also corrected the Role ID resolution prior to calling `findLeader()`.

In addition to those changes, I removed the `withReleaseOnCancel()` setting from the Leader Elector Builder. This was a more recent addition to the Kubernetes Client library implementation. The purpose of the setting is to update the Lease will a null holder identity, prompting nodes to attempt lease renewal. For the purpose of NiFi clustering, this behavior does not seem necessary, as NiFi nodes will proceed with attempting to update and obtain a lease lock. Removing the release on cancel setting avoids the error shown above while allowing standard lease lock update attempts to proceed.

These changes resulted in consistent behavior with various disconnect and reconnect attempts.

markap14 commented:
Thanks for the latest updated @exceptionfactory . Ran into another issue when testing, unfortunately.
I have a statefulset that had 3 replicas. `nifi-1` was both the primary node and the coordinator.
I then scaled the statefulset to 0.
This didn't expire the lease though.:
```
mpayne@cs-654103601966-default:~$ k get leases
NAME                  HOLDER             AGE
cluster-coordinator   nifi-1.nifi:4423   63m
primary-node          nifi-1.nifi:4423   62m
```

Even after I waited over an hour the lease remains there. If I look at it:
```
mpayne@cs-654103601966-default:~$ k get lease cluster-coordinator -o yaml
apiVersion: coordination.k8s.io/v1
kind: Lease
metadata:
  creationTimestamp: "2023-01-05T21:06:29Z"
  name: cluster-coordinator
  namespace: nifi
  resourceVersion: "252479"
  uid: 7e5d05d1-3b20-426d-8822-5cff92eb183f
spec:
  acquireTime: "2023-01-05T22:03:17.355642Z"
  holderIdentity: nifi-1.nifi:4423
  leaseDurationSeconds: 15
  leaseTransitions: 2
  renewTime: "2023-01-05T22:04:13.480562Z"

mpayne@cs-654103601966-default:~$ date
Thu 05 Jan 2023 10:11:34 PM UTC
```

We can see here that date is well past the renewTime. (10:11:34 PM = 22:11:34 PM vs 22:04:13 as the renew time).
So the least appears to remain, and the new node, `nifi-0` cannot proceed:

```
2023-01-05 22:09:37,513 INFO [main] o.a.n.c.p.AbstractNodeProtocolSender Cluster Coordinator is located at nifi-1.nifi:4423. Will send Cluster Connection Request to this address
2023-01-05 22:09:37,535 WARN [main] o.a.nifi.controller.StandardFlowService Failed to connect to cluster due to: org.apache.nifi.cluster.protocol.ProtocolException: Failed to create socket to nifi-1.nifi:4423 due to: java.net.UnknownHostException: nifi-1.nifi
2023-01-05 22:09:42,550 INFO [main] o.a.n.c.c.n.LeaderElectionNodeProtocolSender Determined that Cluster Coordinator is located at nifi-1.nifi:4423; will use this address for sending heartbeat messages
2023-01-05 22:09:42,550 INFO [main] o.a.n.c.p.AbstractNodeProtocolSender Cluster Coordinator is located at nifi-1.nifi:4423. Will send Cluster Connection Request to this address
2023-01-05 22:09:42,550 WARN [main] o.a.nifi.controller.StandardFlowService Failed to connect to cluster due to: org.apache.nifi.cluster.protocol.ProtocolException: Failed to create socket to nifi-1.nifi:4423 due to: java.net.UnknownHostException: nifi-1.nifi
```

As soon as I delete the lease (`k delete lease cluster-coordinator`) all works as expected.
But we obviously can't have users manually deleting the lease all the time.
Not sure if this is the intended behavior, and we should be ignoring the lease if the renewTime has expired? Or is it because we don't actually participate in the leader election on startup since there appears to already be an elected leader?
Either way, we need to make sure that we can properly handle this condition, where the lease points to a node that is no longer part of the cluster

exceptionfactory commented:
Good find on the expired Lease handling @markap14, thanks for the additional testing.

The implementation of `StandardLeaderElectionCommandProvider.findLeader()` was always returning the last Lease Holder Identity, regardless of whether the Lease had expired. This broke the contract of `LeaderElectionManager.getLeader()` which callers expected to return a `null` when there is no leader for the role requested. I updated `findLeader()` to return `Optional<String>`, and check the expiration of the Lease so that `findLeader` returns `Optional.empty()` when the Lease has expired. The expiration is determined based on `renewTime` plus `leaseDurationSeconds`, following the definition of those fields in the [Kubernetes LeaseSpec](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/lease-v1/#LeaseSpec).

To make the implementation contract clearer, I also changed the return of `LeaderElectionManager.getLeader()` from `String` to `Optional<String>`. Although the method documentation noted that `null` indicates no leader is elected, changing the return to use `Optional` clarifies that detail without requiring too many adjustments. This seemed like a good time to make the change as part of promoting `LeaderElectionManager` to an extension component interface.

Following those changes, shutting down all nodes, then starting up a single node that was not previously the Cluster Coordinator or Primary Node worked as expected. The single node was able to become the leader for those roles.

It is worth noting that the Lease objects continue to live in the Kubernetes namespace even if all nodes have been shutdown for an extended period. The `renewTime` indicates the last update, and the corrected logic will now update existing values as needed.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/StandardNiFiServiceFacade.java:
```diff
@@ -6261,8 +6261,8 @@ private Set<String> getRoles(final NodeIdentifier nodeId) {
         for (final String roleName : ClusterRoles.getAllRoles()) {
-            final String leader = leaderElectionManager.getLeader(roleName);
-            if (leader == null) {
+            final Optional<String> leader = leaderElectionManager.getLeader(roleName);
```
The change to `Optional<String>` here means that below, in line 6269, we are calling `leader.equals(nodeAddress)` which compares a `String` to an `Optional<String>` - need to ensure that we call `get()` first. As-is, the cluster page shows the nodes are connected but doesn't show which is Cluster Coordinator and which is primary node (though the nodes do appear to function in those roles properly).

exceptionfactory replied:
Thanks for catching that, I missed pushing that change through to that comparison. Will push a correction.

michael81877 commented:
It looks like a `ConfigMap` is created for each component that requires state management. Do you have any concern about the size of state that may be written to any one `ConfigMap`?

The [maximum size of a ConfigMap is 1MiB](https://kubernetes.io/docs/concepts/configuration/configmap/#:~:text=The%20data%20stored%20in%20a,separate%20database%20or%20file%20service.). Do you anticipate any compatibility issues where a use case would work with Zookeeper but would not work with ConfigMaps?

exceptionfactory commented:
> It looks like a `ConfigMap` is created for each component that requires state management. Do you have any concern about the size of state (in bytes) that may be written to any one `ConfigMap`?
> 
> The [maximum size of a ConfigMap is 1MiB](https://kubernetes.io/docs/concepts/configuration/configmap/#:~:text=The%20data%20stored%20in%20a,separate%20database%20or%20file%20service.). Do you anticipate any compatibility issues where a use case would work with Zookeeper but would not work with ConfigMaps?

Thanks for the feedback @michael81877. The standard ZooKeeper node storage is also [limited to 1 MB](https://zookeeper.apache.org/doc/r3.8.0/apidocs/zookeeper-server/org/apache/zookeeper/ZooKeeper.html#create-java.lang.String-byte:A-java.util.List-org.apache.zookeeper.CreateMode-), so the standard size limits of a Kubernetes ConfigMap align well with current behavior.

With that background and similar size limitations, I do not expect any compatibility issues based on size limitations. It is possible that other behavioral nuances could surface as a result of these alternative State Management and Leader Election implementations.

exceptionfactory commented:
@markap14 I rebased the branch to the current `main` so that all modules now reference the current 2.0.0-SNAPSHOT version.

bbende reviewed (approved):
Ran this in k8s and tested all of the basic clustering scenarios like disconnecting/connecting nodes, killing nodes, storing cluster state, etc., everything looks good, nice work! Going to merge this.
